models/ROMS_rutgers/model_mod.f90

```diff
@@ -0,0 +1,1716 @@
+call nc_close_file(ncid, routine)
+
+end function read_model_time
+
+
+!-----------------------------------------------------------------------
+! Set the desired minimum model advance time. This is generally NOT the
+! dynamical timestep of the model, but rather the shortest forecast length
+! you are willing to make. This impacts how frequently the observations
+! may be assimilated.
+
+function set_model_time_step()
+
+type(time_type) :: set_model_time_step
+
+! assimilation_period_seconds, assimilation_period_days are from the namelist
+set_model_time_step = set_time(assimilation_period_seconds, assimilation_period_days)
+
+end function set_model_time_step
+
+
+!-----------------------------------------------------------------------
+! Read the grid dimensions from the ROMS grid netcdf file.
+! By reading the dimensions first, we can use them in variable
+! declarations later - which is faster than using allocatable arrays.
+
+subroutine get_grid_dimensions()
+
+character(len=*), parameter :: routine = 'get_grid_dimensions'
+
+integer :: ncid
+integer :: Nxi_rho, Nxi_u, Nxi_v
+integer :: Neta_rho, Neta_u, Neta_v
+integer :: Ns_rho, Ns_w
+
+! Read the (static) grid dimensions from the ROMS grid file.
+ncid = nc_open_file_readonly(roms_filename, routine)
+
+Nxi_rho  = nc_get_dimension_size(ncid, 'xi_rho',  routine)
+Nxi_u    = nc_get_dimension_size(ncid, 'xi_u',    routine)
+Nxi_v    = nc_get_dimension_size(ncid, 'xi_v',    routine)
+Neta_rho = nc_get_dimension_size(ncid, 'eta_rho', routine)
+Neta_u   = nc_get_dimension_size(ncid, 'eta_u',   routine)
+Neta_v   = nc_get_dimension_size(ncid, 'eta_v',   routine)
+
+! Read the vertical dimensions 
+Ns_rho   = nc_get_dimension_size(ncid, 's_rho', routine)
+Ns_w     = nc_get_dimension_size(ncid, 's_w'  , routine)
+
+call nc_close_file(ncid, routine)
+
+Nx = Nxi_rho  ! Setting the nominal value of the 'global' variables
+Ny = Neta_rho ! Setting the nominal value of the 'global' variables
+Nz = Ns_rho   ! Setting the nominal value of the 'global' variables
+
+Nu = Nxi_u
+Nv = Neta_v
+Nw = Ns_w
+
+end subroutine get_grid_dimensions
+
+
+!-----------------------------------------------------------------------
+! Read the actual grid values from the ROMS netcdf file.
+! We might need to adapt this to allow for leapfrog time stepping
+! The assumption is that we're reading from history files 
+! which only have one time-level. 
+
+subroutine get_grid()
+
+integer                     :: ncid
+character(len=*), parameter :: routine = 'get_grid'
+
+real(r8), allocatable       :: zeta(:,:), zeta_lf(:,:,:) 
+real(r8), allocatable       :: mask(:,:)                 ! Land mask: 0 land & 1 water
+real(r8)                    :: Zm = -20000.0_r8          ! a masking factor to account for land
+
+allocate(ULAT(Nu, Ny), ULON(Nu, Ny), UMSK(Nu, Ny), UDEP(Nu, Ny, Nz))
+allocate(VLAT(Nx, Nv), VLON(Nx, Nv), VMSK(Nx, Nv), VDEP(Nx, Nv, Nz))
+allocate(TLAT(Nx, Ny), TLON(Nx, Ny), TMSK(Nx, Ny), TDEP(Nx, Ny, Nz))
```

Comment:
These arrays will get large for high-res models. @todo may need to distribute this data in the future.

Reply:
We should talk about this. I have ideas and concerns about depth for ROMS ..